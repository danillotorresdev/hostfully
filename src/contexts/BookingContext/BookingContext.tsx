import { ReactNode, createContext, useContext, useState } from "react";
import { z } from "zod";

const PassengerSchema = z.object({
  adults: z.number(),
  children: z.number(),
});

const BookingSchema = z.object({
  id: z.string(),
  startDate: z.string(),
  endDate: z.string(),
  origin: z.string(),
  destination: z.string(),
  passengers: PassengerSchema,
});

type Booking = z.infer<typeof BookingSchema>;

const BookingContext = createContext<{
  bookings: Booking[];
  createBooking: (booking: Booking) => void;
  deleteBooking: (bookingId: string) => void;
  validateBookingCreation: (booking: Booking) => boolean;
  validateBookingUpdate: (booking: Booking) => boolean;
  updateBooking: (booking: Booking) => void;
}>({
  bookings: [],
  createBooking: () => {},
  deleteBooking: () => {},
  validateBookingCreation: () => false,
  validateBookingUpdate: () => false,
  updateBooking: () => {},
});

export const useBookingContext = () => useContext(BookingContext);

type BookingProviderProps = {
  children: ReactNode;
};

export const BookingProvider = ({ children }: BookingProviderProps) => {
  const [bookings, setBookings] = useState<Booking[]>([
    {
      id: "1",
      origin: "Cairo",
      destination: "Great Pyramid of Giza",
      startDate: "2024-01-11T03:00:00.000Z",
      endDate: "2024-01-15T03:00:00.000Z",
      passengers: {
        adults: 1,
        children: 0,
      },
    },
    {
      id: "2",
      origin: "Bodrum",
      destination: "Mausoleum at Halicarnassus",
      startDate: "2024-01-11T03:00:00.000Z",
      endDate: "2024-01-15T03:00:00.000Z",
      passengers: {
        adults: 2,
        children: 1,
      },
    },
  ]);

  const createBooking = (booking: Booking) => {
    const validBookingData = BookingSchema.parse(booking);
    setBookings([validBookingData, ...bookings]);
  };

  const deleteBooking = (bookingId: string) => {
    setBookings(bookings.filter((booking) => booking.id !== bookingId));
  };

  const updateBooking = (booking: Booking) => {
    const validBookingData = BookingSchema.parse(booking);

    const foundBooking = bookings.find(
      (booking) => booking.id === validBookingData.id,
    );

    if (!foundBooking) {
      return;
    }

    const updatedBookings = bookings.map((booking) =>
      booking.id === validBookingData.id ? validBookingData : booking,
    );

    setBookings(updatedBookings);
  };

  const validateBookingCreation = (booking: Booking) => {
    const newBookingStartDate = new Date(booking.startDate);
    const newBookingEndDate = new Date(booking.endDate);

    for (const existingBooking of bookings) {
      const existingBookingStartDate = new Date(existingBooking.startDate);
      const existingBookingEndDate = new Date(existingBooking.endDate);
      if (
        (newBookingStartDate >= existingBookingStartDate &&
          newBookingStartDate <= existingBookingEndDate) ||
        (newBookingEndDate >= existingBookingStartDate &&
          newBookingEndDate <= existingBookingEndDate)
      ) {
        return false;
      }
    }

    return true;
  };

  const validateBookingUpdate = (booking: Booking) => {
    const newBookingStartDate = new Date(booking.startDate);
    const newBookingEndDate = new Date(booking.endDate);

    const selectedBooking = bookings.find(
      (existingBooking) => existingBooking.id === booking.id,
    );

    if (!selectedBooking) {
      return false;
    }

    const existingBookingStartDate = new Date(selectedBooking.startDate);
    const existingBookingEndDate = new Date(selectedBooking.endDate);

    if (
      newBookingStartDate.getTime() === existingBookingStartDate.getTime() &&
      newBookingEndDate.getTime() === existingBookingEndDate.getTime()
    ) {
      return true;
    }

    for (const existingBooking of bookings) {
      const existingBookingStartDate = new Date(existingBooking.startDate);
      const existingBookingEndDate = new Date(existingBooking.endDate);
      if (
        (newBookingStartDate >= existingBookingStartDate &&
          newBookingStartDate <= existingBookingEndDate) ||
        (newBookingEndDate >= existingBookingStartDate &&
          newBookingEndDate <= existingBookingEndDate)
      ) {
        return false;
      }
    }

    return true;
  };

  return (
    <BookingContext.Provider
      value={{
        bookings,
        createBooking,
        deleteBooking,
        updateBooking,
        validateBookingCreation,
        validateBookingUpdate,
      }}
    >
      {children}
    </BookingContext.Provider>
  );
};
