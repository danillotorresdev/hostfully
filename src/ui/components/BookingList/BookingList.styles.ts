import { styled } from "@linaria/react";

export const Container = styled.div`
  border-top: 1px solid #e0e0e0;
  padding-top: 40px;

  @media (min-width: 768px) {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  @media (min-width: 1024px) {
    display: inherit;
    flex-direction: inherit;
    justify-content: inherit;
    align-items: inherit;
  }
`;

export const Title = styled.h2`
  color: #333;
  font-size: 24px;
  margin: 0;

  @media (min-width: 768px) {
    margin-bottom: 20px;
  }
`;
