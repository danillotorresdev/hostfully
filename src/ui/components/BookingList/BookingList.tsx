import { useBookingContext } from "@/contexts/BookingContext/BookingContext";
import * as S from "./BookingList.styles";
import { Itinerary } from "@/ui/components/Itinerary/Itinerary";

export function BookingList() {
  const { bookings, deleteBooking } = useBookingContext();

  if (bookings.length === 0) {
    return (
      <S.Container>
        <h2>No bookings yet</h2>
      </S.Container>
    );
  }

  return (
    <S.Container>
      <S.Title>Bookings</S.Title>
      {bookings.map((booking) => (
        <Itinerary
          key={booking.id}
          booking={booking}
          deleteBooking={deleteBooking}
        />
      ))}
    </S.Container>
  );
}
