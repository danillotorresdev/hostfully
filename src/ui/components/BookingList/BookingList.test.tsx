import { BookingList } from "./BookingList";
import { describe, expect, it } from "vitest";
import * as BookingContext from "@/contexts/BookingContext/BookingContext";

import { render, screen, fireEvent } from "@/utils/settings/tests/utilities";

describe("BookingList", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <BookingContext.BookingProvider>
        <BookingList />
      </BookingContext.BookingProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  
  describe("No bookings", () => {
    it("should show no bookings message and delete booking", async () => {
      render(
        <BookingContext.BookingProvider>
          <BookingList />
        </BookingContext.BookingProvider>,
      );

      const deleteButton = screen.getAllByTitle("Delete booking");

      const button1 = deleteButton[0] as HTMLButtonElement;
      const button2 = deleteButton[1] as HTMLButtonElement;

      fireEvent.click(button1);
      fireEvent.click(button2);

      expect(screen.getByText("No bookings yet")).toBeDefined();
    });
  });

  describe("With bookings", () => {
    it("should show bookings", () => {
      render(
        <BookingContext.BookingProvider>
          <BookingList />
        </BookingContext.BookingProvider>,
      );

      expect(screen.getByText("Bookings")).toBeDefined();
    });
  });
});
