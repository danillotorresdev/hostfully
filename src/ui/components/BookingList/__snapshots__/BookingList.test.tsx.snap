// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookingList > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-0 crlin7j"
  >
    <h2
      class="mocked-styled-1 t1bjt9tu"
    >
      Bookings
    </h2>
    <div
      class="mocked-styled-2 csr9ikx"
    >
      <div
        class="mocked-styled-3 cx2fecm"
      >
        <span
          class="mocked-styled-5 ta36wq1"
        >
          Great Pyramid of Giza
        </span>
        <span
          class="mocked-styled-6 s1a79fbf"
        >
          Booking for 1 Adult   | Jan 11, 2024 to Jan 15, 2024
        </span>
        <div
          class="mocked-styled-4 c16asigh"
        >
          <button
            class="mocked-styled-10 bb8cb6j"
            style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
            title="Edit booking"
          >
            <svg
              class="svg svg-edit"
              data-testid="icon-svg"
              style="width: 24px; height: 24px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#edit"
              />
            </svg>
          </button>
          <button
            class="mocked-styled-10 bb8cb6j"
            style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
            title="Delete booking"
          >
            <svg
              class="svg svg-trash"
              data-testid="icon-svg"
              style="width: 24px; height: 24px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#trash"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="mocked-styled-8 c6r2qql"
      >
        <div
          class="mocked-styled-11 ix3zkq7"
        >
          <div
            class="mocked-styled-12 c1n5q6aa"
          >
            <div>
              <svg
                class="svg svg-departure"
                data-testid="icon-svg"
                style="width: 16px; height: 16px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#departure"
                />
              </svg>
               
              <b>
                Departure
              </b>
            </div>
            <span
              class="mocked-styled-13 dsggpze"
            >
              Thu, Jan 11, 2024
            </span>
          </div>
          <div
            class="mocked-styled-14 c1yloby"
          >
            <span
              title="Cairo"
            >
              Cairo 
            </span>
            <svg
              class="svg svg-arrow-foward"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#arrow-foward"
              />
            </svg>
             
            <span
              title="Great Pyramid of Giza"
            >
              Great Pyramid of Giz...
            </span>
          </div>
        </div>
        <div
          class="mocked-styled-11 ix3zkq7"
        >
          <div
            class="mocked-styled-12 c1n5q6aa"
          >
            <div>
              <svg
                class="svg svg-arrival"
                data-testid="icon-svg"
                style="width: 16px; height: 16px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#arrival"
                />
              </svg>
               
              <b>
                Return
              </b>
            </div>
            <span
              class="mocked-styled-13 dsggpze"
            >
              Mon, Jan 15, 2024
            </span>
          </div>
          <div
            class="mocked-styled-14 c1yloby"
          >
            <span
              title="Cairo"
            >
              Cairo 
            </span>
            <svg
              class="svg svg-arrow-back"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#arrow-back"
              />
            </svg>
             
            <span
              title="Great Pyramid of Giza"
            >
              Great Pyramid of Giz...
            </span>
          </div>
        </div>
      </div>
      <div
        class="mocked-styled-9 c1wkgrf8"
      >
        <svg
          class="svg svg-members"
          data-testid="icon-svg"
          style="width: 16px; height: 16px;"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="/assets/sprite.svg#members"
          />
        </svg>
        1 Adult 
      </div>
    </div>
    <div
      class="mocked-styled-2 csr9ikx"
    >
      <div
        class="mocked-styled-3 cx2fecm"
      >
        <span
          class="mocked-styled-5 ta36wq1"
        >
          Mausoleum at Halicarnassus
        </span>
        <span
          class="mocked-styled-6 s1a79fbf"
        >
          Booking for 2 Adults and 1 Child  | Jan 11, 2024 to Jan 15, 2024
        </span>
        <div
          class="mocked-styled-4 c16asigh"
        >
          <button
            class="mocked-styled-10 bb8cb6j"
            style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
            title="Edit booking"
          >
            <svg
              class="svg svg-edit"
              data-testid="icon-svg"
              style="width: 24px; height: 24px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#edit"
              />
            </svg>
          </button>
          <button
            class="mocked-styled-10 bb8cb6j"
            style="--bb8cb6j-0: 40px; --bb8cb6j-1: 14px; --bb8cb6j-2: 0; --bb8cb6j-3: transparent; --bb8cb6j-4: #2D2AA5; --bb8cb6j-5: none;"
            title="Delete booking"
          >
            <svg
              class="svg svg-trash"
              data-testid="icon-svg"
              style="width: 24px; height: 24px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#trash"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="mocked-styled-8 c6r2qql"
      >
        <div
          class="mocked-styled-11 ix3zkq7"
        >
          <div
            class="mocked-styled-12 c1n5q6aa"
          >
            <div>
              <svg
                class="svg svg-departure"
                data-testid="icon-svg"
                style="width: 16px; height: 16px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#departure"
                />
              </svg>
               
              <b>
                Departure
              </b>
            </div>
            <span
              class="mocked-styled-13 dsggpze"
            >
              Thu, Jan 11, 2024
            </span>
          </div>
          <div
            class="mocked-styled-14 c1yloby"
          >
            <span
              title="Bodrum"
            >
              Bodrum 
            </span>
            <svg
              class="svg svg-arrow-foward"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#arrow-foward"
              />
            </svg>
             
            <span
              title="Mausoleum at Halicarnassus"
            >
              Mausoleum at Halicar...
            </span>
          </div>
        </div>
        <div
          class="mocked-styled-11 ix3zkq7"
        >
          <div
            class="mocked-styled-12 c1n5q6aa"
          >
            <div>
              <svg
                class="svg svg-arrival"
                data-testid="icon-svg"
                style="width: 16px; height: 16px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#arrival"
                />
              </svg>
               
              <b>
                Return
              </b>
            </div>
            <span
              class="mocked-styled-13 dsggpze"
            >
              Mon, Jan 15, 2024
            </span>
          </div>
          <div
            class="mocked-styled-14 c1yloby"
          >
            <span
              title="Bodrum"
            >
              Bodrum 
            </span>
            <svg
              class="svg svg-arrow-back"
              data-testid="icon-svg"
              style="width: 16px; height: 16px;"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#arrow-back"
              />
            </svg>
             
            <span
              title="Mausoleum at Halicarnassus"
            >
              Mausoleum at Halicar...
            </span>
          </div>
        </div>
      </div>
      <div
        class="mocked-styled-9 c1wkgrf8"
      >
        <svg
          class="svg svg-members"
          data-testid="icon-svg"
          style="width: 16px; height: 16px;"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="/assets/sprite.svg#members"
          />
        </svg>
        2 Adults and 1 Child
      </div>
    </div>
  </div>
</DocumentFragment>
`;
